tests/msc2716_test.go

```diff
@@ -0,0 +1,131 @@
+// +build msc2716
+
+// This file contains tests for incrementally importing history to an existing room,
+// a currently experimental feature defined by MSC2716, which you can read here:
+// https://github.com/matrix-org/matrix-doc/pull/2716
+
+package tests
+
+import (
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/sirupsen/logrus"
+	"github.com/tidwall/gjson"
+)
+
+// Test that the m.room.create and m.room.member events for a room we created comes down /sync
+func TestBackfillingHistory(t *testing.T) {
+	deployment := Deploy(t, "rooms_state", b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	userID := "@alice:hs1"
+	alice := deployment.Client(t, "hs1", userID)
+	roomID := alice.CreateRoom(t, struct{}{})
+
+	// eventA
+	eventA := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message A",
+		},
+	})
+
+	insertTime := time.Now()
+	insertOriginServerTs := uint64(insertTime.UnixNano() / 1000000)
+
+	// wait 3ms to ensure that the timestamp changes enough intervals for each message we try to insert later
+	time.Sleep(3 * time.Millisecond)
+
+	// eventB
+	alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message B",
+		},
+	})
+	// eventC
+	alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message C",
+		},
+	})
+
+	// event1
+	event1 := alice.SendEvent(t, roomID, b.Event{
+		Type: "m.room.message",
+		PrevEvents: []string{
+			eventA,
+		},
+		OriginServerTS: insertOriginServerTs,
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message 1",
+		},
+	})
+
+	// event2
+	event2 := alice.SendEvent(t, roomID, b.Event{
+		Type: "m.room.message",
+		PrevEvents: []string{
+			event1,
+		},
+		OriginServerTS: insertOriginServerTs + 1,
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message 2",
+		},
+	})
+
+	// event3
+	alice.SendEvent(t, roomID, b.Event{
+		Type: "m.room.message",
+		PrevEvents: []string{
+			event2,
+		},
+		OriginServerTS: insertOriginServerTs + 2,
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message 3",
+		},
+	})
+
+	res := alice.MustDoRaw(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, nil, "application/json", url.Values{
```

Comment:
The current code will return the events in the following order. I expect them to be in A <- 1 <- 2 <- 3 <- B <- C where I am inserting `1-3` between `A` and `B`.

![](https://user-images.githubusercontent.com/558581/106347004-b3fd7400-6280-11eb-9beb-76b4d4a2dca1.png)



<details>

<summary>Messages result</summary>

```json
{
  "chunk": [
    {
      "type": "m.room.message",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 3",
        "msgtype": "m.text"
      },
      "origin_server_ts": 1611980454461,
      "unsigned": {
        "age": 30
      },
      "event_id": "$1U0RP8BvQ7GSoRkbURyCjlsyGuG7D1GMEDzLiku0WfA",
      "user_id": "@alice:hs1",
      "age": 30
    },
    {
      "type": "m.room.message",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 2",
        "msgtype": "m.text"
      },
      "origin_server_ts": 1611980454460,
      "unsigned": {
        "age": 54
      },
      "event_id": "$_izxcHvKcX86vdvw1H5nipn9EOjbnCpV8vjAGiCob14",
      "user_id": "@alice:hs1",
      "age": 54
    },
    {
      "type": "m.room.message",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message C",
        "msgtype": "m.text"
      },
      "origin_server_ts": 1611980454503,
      "unsigned": {
        "age": 110
      },
      "event_id": "$aVRthsR9iJxT4E_dv5SAXfSApH8CmYnod80XVF0SFPM",
      "user_id": "@alice:hs1",
      "age": 110
    },
    {
      "type": "m.room.message",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 1",
        "msgtype": "m.text"
      },
      "origin_server_ts": 1611980454459,
      "unsigned": {
        "age": 74
      },
      "event_id": "$Ulkv0PVlHnGiOpSOKJEkV3Y6d2yJam1O5IlWiT_iMIM",
      "user_id": "@alice:hs1",
      "age": 74
    },
    {
      "type": "m.room.message",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message B",
        "msgtype": "m.text"
      },
      "origin_server_ts": 1611980454466,
      "unsigned": {
        "age": 147
      },
      "event_id": "$-ltydi9ucXLipXHTx6Cy2SNKbBbv57yViEqmX2WmwYM",
      "user_id": "@alice:hs1",
      "age": 147
    },
    {
      "type": "m.room.message",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message A",
        "msgtype": "m.text"
      },
      "origin_server_ts": 1611980454402,
      "unsigned": {
        "age": 211
      },
      "event_id": "$7bcRBvXtCpOH5RWIFhV2PNirj6Ygo-8Ds3GB5CxxwWY",
      "user_id": "@alice:hs1",
      "age": 211
    },
    {
      "type": "m.room.history_visibility",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "state_key": "",
      "origin_server_ts": 1611980454361,
      "unsigned": {
        "age": 252
      },
      "event_id": "$Pnkj7_XyohjRqbghpHf0bz7voe3WQ31Jk8AYP-5myQI",
      "user_id": "@alice:hs1",
      "age": 252
    },
    {
      "type": "m.room.join_rules",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "join_rule": "public"
      },
      "state_key": "",
      "origin_server_ts": 1611980454340,
      "unsigned": {
        "age": 273
      },
      "event_id": "$6CZ9mMoiMhgxKPB9jxijJM0mHFcf-G1a3s1ntFYxl2A",
      "user_id": "@alice:hs1",
      "age": 273
    },
    {
      "type": "m.room.power_levels",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "users": {
          "@alice:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "state_key": "",
      "origin_server_ts": 1611980454319,
      "unsigned": {
        "age": 294
      },
      "event_id": "$U4uTUWi65wAhyWy4gGP22MdFbD0duxIj6CsrggfF5w0",
      "user_id": "@alice:hs1",
      "age": 294
    },
    {
      "type": "m.room.member",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "state_key": "@alice:hs1",
      "origin_server_ts": 1611980454292,
      "unsigned": {
        "age": 321
      },
      "event_id": "$mDL-UHiXQRtnQwY3QKSfLL2DUikkPSIsduRXT9ltlQg",
      "user_id": "@alice:hs1",
      "age": 321
    },
    {
      "type": "m.room.create",
      "room_id": "!PSAYnIBCViJErRsSbp:hs1",
      "sender": "@alice:hs1",
      "content": {
        "room_version": "6",
        "creator": "@alice:hs1"
      },
      "state_key": "",
      "origin_server_ts": 1611980454251,
      "unsigned": {
        "age": 362
      },
      "event_id": "$2iNyttb3Yr8iXbwKNPdr355WAlEzKmq16KGZoTRjgrE",
      "user_id": "@alice:hs1",
      "age": 362
    }
  ],
  "start": "s12_0_0_0_0_0_0_0_0",
  "end": "t1-1_0_0_0_0_0_0_0_0"
}
```

</details>

Reply:
Maybe I can get a good result out of [`contrib/graph/graph.py`](https://github.com/matrix-org/synapse/blob/f2c1560eca1e2160087a280261ca78d0708ad721/contrib/graph/graph.py) (there is also `graph2` and `graph3`)

Reply:
Here is a visual representation which looks good. I haven't tried healing the state and tying it back in.

TARDIS edits to make this happen are in https://github.com/matrix-org/tardis/pull/1

![](https://user-images.githubusercontent.com/558581/106574040-05fffd00-6500-11eb-94b8-e71be813c2d8.png)


[`GET /_matrix/client/r0/rooms/{roomId}/context/{eventId}`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-context-eventid)
<details>

<summary>Response:</summary>

```json
{
  "events_before": [
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$RcUz4kTgxExqtDWZy9GejprO_tpY2eUTibCx3aHqYvY"
      ],
      "type": "m.room.message",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 3",
        "msgtype": "m.text"
      },
      "depth": 9,
      "prev_state": [],
      "origin_server_ts": 1612227218616,
      "origin": "hs1",
      "hashes": {
        "sha256": "8mP02Jd9HS0oO+dn4R7kujGCjZFqmS5wPI/uLSMstOA"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "chxruiqU1TftetfZOSXvk94zoXsn/PibE1VI3wnmiZ8mdXPRfgXUmWoB9NW40T9j4TLodMwk0DX3XClLdEn+DA"
        }
      },
      "unsigned": {
        "age": 124
      },
      "event_id": "$_bLBHOgPgDihh5E7NKLxsyKCnkxxdN-l8ueh8tdebJM",
      "user_id": "@alice:hs1",
      "age": 124
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$AN5Tli5_7sCvii2Cv563bkw3PLwtWYmV0CvRqSku9m4"
      ],
      "type": "m.room.message",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 2",
        "msgtype": "m.text"
      },
      "depth": 8,
      "prev_state": [],
      "origin_server_ts": 1612227218615,
      "origin": "hs1",
      "hashes": {
        "sha256": "FFnpBdhRUmhCNAE8y289dTrVh86mClDWbB6T7DHb3+A"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "geuJQt1esPNOyuL9hSdDnTl1CfLlzPSG0/2UkfIgc0WRQr291XhhRrQ3r7+Xo7c1tzaiD4ShlMQG3jSbdGgjAA"
        }
      },
      "unsigned": {
        "age": 156
      },
      "event_id": "$RcUz4kTgxExqtDWZy9GejprO_tpY2eUTibCx3aHqYvY",
      "user_id": "@alice:hs1",
      "age": 156
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$VF7hNqLEjuqwDwHpxIy9MFKEBJE-5NXIjyrRxzEQUOQ"
      ],
      "type": "m.room.message",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message C",
        "msgtype": "m.text"
      },
      "depth": 8,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612227218693,
      "hashes": {
        "sha256": "ZRNuAeyBIwotdACQmp2C8DNP3Bj4+7oF5uuB7NY0Hag"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "Q1nGVVCdya/lbDwON31NYfVUyQBU/r9kiF+fhwEPJzpxVf0zGgY6QdRaH7t95LfcIGusLj3uFmTps2dLfj3YBg"
        }
      },
      "unsigned": {
        "age": 235
      },
      "event_id": "$92x9aKGxaxt5aVhBul0IySe6vk_IKNRCrgW05rqn3aY",
      "user_id": "@alice:hs1",
      "age": 235
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$tUAkGmH62FbpPn8S8siFE0KT3N8WXZmpAh7hjBGUgPY"
      ],
      "type": "m.room.message",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 1",
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin_server_ts": 1612227218614,
      "origin": "hs1",
      "hashes": {
        "sha256": "CgH6A+GkC19bsc0uxDjwaWkGdEJQJBYE69IeUI1CzKg"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "bu74sLCmyiyaRA61Y7skMLtawBZO/Snv+mYZ82dhKuy1He6gOwsw+NzOdPOtwx8ur0KY/PqoZacHivj/R04TAg"
        }
      },
      "unsigned": {
        "age": 182
      },
      "event_id": "$AN5Tli5_7sCvii2Cv563bkw3PLwtWYmV0CvRqSku9m4",
      "user_id": "@alice:hs1",
      "age": 182
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$tUAkGmH62FbpPn8S8siFE0KT3N8WXZmpAh7hjBGUgPY"
      ],
      "type": "m.room.message",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message B",
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612227218634,
      "hashes": {
        "sha256": "SGD23OPuNtV/sda/2T3k37wRLiEZ6TOOeuL4IOHDZl0"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "5adPRqXmZ9E7JdIiNw5lcjOdRcUg7EcY6rRr0H+2fjz8W+aqx+/+ipduNjZxiqVvesxkiTbRNpho5ZBVbRXWAg"
        }
      },
      "unsigned": {
        "age": 294
      },
      "event_id": "$VF7hNqLEjuqwDwHpxIy9MFKEBJE-5NXIjyrRxzEQUOQ",
      "user_id": "@alice:hs1",
      "age": 294
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$xCtSXTydyn3uatfvyl_C8-Iq2wxgRpC5bSxNc9191-g"
      ],
      "type": "m.room.message",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message A",
        "msgtype": "m.text"
      },
      "depth": 6,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612227218542,
      "hashes": {
        "sha256": "TN/hfy3hdGIvIdOY4LpAOoCa421lGPhacs+lOQhZ7Q0"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "ZZW7UvS4fgD1YhOXtrn/2oIZ0/4WV/Y0F4QP9v5TbM0aG0MP5wz8g38vBVdkv7Rgl1PuhNHntDo1u3K9cvSUDQ"
        }
      },
      "unsigned": {
        "age": 386
      },
      "event_id": "$tUAkGmH62FbpPn8S8siFE0KT3N8WXZmpAh7hjBGUgPY",
      "user_id": "@alice:hs1",
      "age": 386
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$xEfMKOL59s7305p_s9Kx0TTuKLWhLHH-qNgygg8zQEw"
      ],
      "type": "m.room.history_visibility",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "depth": 5,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218481,
      "hashes": {
        "sha256": "/lSrm4/Y4CWDGsVRZnFSbiP3YRNivFR7TQARVtMXQQU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "mjaE56SnX3A3C4soy48lKevOtjLxGqv4gxavCg0tsxfSCMhvVlzbYMofAX23K7Aq8Tg4cF3bf8HIVMkeNsLMAA"
        }
      },
      "unsigned": {
        "age": 447
      },
      "event_id": "$xCtSXTydyn3uatfvyl_C8-Iq2wxgRpC5bSxNc9191-g",
      "user_id": "@alice:hs1",
      "age": 447
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "type": "m.room.join_rules",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "join_rule": "public"
      },
      "depth": 4,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218448,
      "hashes": {
        "sha256": "3+iuQliDoWsIrkY/tK4EhD+aoeW3nEoL9QSe4c197OU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "Wk7NtPF0+Urd5NYUsaQwrGuOYs7Pio2bztu6TVXDTjgo6lIBtQNwWDte1lNzIP5TK4WEFofR7sl8NoYE9DYZAA"
        }
      },
      "unsigned": {
        "age": 480
      },
      "event_id": "$xEfMKOL59s7305p_s9Kx0TTuKLWhLHH-qNgygg8zQEw",
      "user_id": "@alice:hs1",
      "age": 480
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY"
      ],
      "prev_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8"
      ],
      "type": "m.room.power_levels",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "users": {
          "@alice:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "depth": 3,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218413,
      "hashes": {
        "sha256": "bp3AcMT21CsKy3JDiQBbghGJuNftEz+HUB3X/XBdocs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "zoGF6psTpZVUP1hvaxc0ElgILnkMcEJToiqcQQwOo8au8EaCw7yyXPn35KGUWn77CU6yup70L4dUa9bVARLyBQ"
        }
      },
      "unsigned": {
        "age": 515
      },
      "event_id": "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk",
      "user_id": "@alice:hs1",
      "age": 515
    },
    {
      "auth_events": [
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY"
      ],
      "prev_events": [
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY"
      ],
      "type": "m.room.member",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "depth": 2,
      "prev_state": [],
      "state_key": "@alice:hs1",
      "origin": "hs1",
      "origin_server_ts": 1612227218374,
      "hashes": {
        "sha256": "Ijm11zXRu0bHwFUSDCULX878nJtVXxSYUWbdYcMSBQs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "guhM3fw+cXOOQ++Qb/wmZJkednj7XnTDqMx84daq3BO8FGznV4kYrpb1PWKFhsCRX3IULSCNq/+0Q0ysdM9FBA"
        }
      },
      "unsigned": {
        "age": 554
      },
      "event_id": "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
      "user_id": "@alice:hs1",
      "age": 554
    },
    {
      "auth_events": [],
      "prev_events": [],
      "type": "m.room.create",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "room_version": "6",
        "creator": "@alice:hs1"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218316,
      "hashes": {
        "sha256": "sxgLivP2OWcQ/52fEu7go5azKZ2Jomf1Vh+MX4eMsW0"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "3J6FB4wBoNwYr1xLRidjXvl8aZ5ujgD16azhb83O/+G5A2A1ONLjYxRgidrI5zCx6s9+9hQFAd0CSgtu41ryBg"
        }
      },
      "unsigned": {
        "age": 612
      },
      "event_id": "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
      "user_id": "@alice:hs1",
      "age": 612
    }
  ],
  "events_after": [],
  "start": "t1-1_0_0_0_0_0_0_0_0",
  "end": "t10-13_0_0_0_0_0_0_0_0",
  "event": {
    "auth_events": [
      "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
      "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
      "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
    ],
    "prev_events": [
      "$_bLBHOgPgDihh5E7NKLxsyKCnkxxdN-l8ueh8tdebJM",
      "$92x9aKGxaxt5aVhBul0IySe6vk_IKNRCrgW05rqn3aY"
    ],
    "type": "m.room.message",
    "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
    "sender": "@alice:hs1",
    "content": {
      "body": "Message *",
      "msgtype": "m.text"
    },
    "depth": 10,
    "prev_state": [],
    "origin": "hs1",
    "origin_server_ts": 1612227218843,
    "hashes": {
      "sha256": "CKDr404T4N78Veiv8h+3XCgm3Wg6Uck+FC9WOx2EAwI"
    },
    "signatures": {
      "hs1": {
        "ed25519:a_BXDh": "ukDEsQR0lzR3PFWOFU6VcuNcwai/cLm+O/9i+eo7QzYKlOTf/E6OLXUAQhY7pbHwhwYcSydP+vshSl9ve560Bw"
      }
    },
    "unsigned": {
      "age": 85
    },
    "event_id": "$l-gc9mYSYrUlIUiGSpRf4J1jU2c8FMjYlP1KUZZX7Xc",
    "user_id": "@alice:hs1",
    "age": 85
  },
  "state": [
    {
      "auth_events": [],
      "prev_events": [],
      "type": "m.room.create",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "room_version": "6",
        "creator": "@alice:hs1"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218316,
      "hashes": {
        "sha256": "sxgLivP2OWcQ/52fEu7go5azKZ2Jomf1Vh+MX4eMsW0"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "3J6FB4wBoNwYr1xLRidjXvl8aZ5ujgD16azhb83O/+G5A2A1ONLjYxRgidrI5zCx6s9+9hQFAd0CSgtu41ryBg"
        }
      },
      "unsigned": {
        "age": 612
      },
      "event_id": "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
      "user_id": "@alice:hs1",
      "age": 612
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY"
      ],
      "prev_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8"
      ],
      "type": "m.room.power_levels",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "users": {
          "@alice:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "depth": 3,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218413,
      "hashes": {
        "sha256": "bp3AcMT21CsKy3JDiQBbghGJuNftEz+HUB3X/XBdocs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "zoGF6psTpZVUP1hvaxc0ElgILnkMcEJToiqcQQwOo8au8EaCw7yyXPn35KGUWn77CU6yup70L4dUa9bVARLyBQ"
        }
      },
      "unsigned": {
        "age": 515
      },
      "event_id": "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk",
      "user_id": "@alice:hs1",
      "age": 515
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "type": "m.room.join_rules",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "join_rule": "public"
      },
      "depth": 4,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218448,
      "hashes": {
        "sha256": "3+iuQliDoWsIrkY/tK4EhD+aoeW3nEoL9QSe4c197OU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "Wk7NtPF0+Urd5NYUsaQwrGuOYs7Pio2bztu6TVXDTjgo6lIBtQNwWDte1lNzIP5TK4WEFofR7sl8NoYE9DYZAA"
        }
      },
      "unsigned": {
        "age": 480
      },
      "event_id": "$xEfMKOL59s7305p_s9Kx0TTuKLWhLHH-qNgygg8zQEw",
      "user_id": "@alice:hs1",
      "age": 480
    },
    {
      "auth_events": [
        "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY",
        "$TvggOE6uk48uXu5b2T7d0q1bfk594k60AKxRbxqucMk"
      ],
      "prev_events": [
        "$xEfMKOL59s7305p_s9Kx0TTuKLWhLHH-qNgygg8zQEw"
      ],
      "type": "m.room.history_visibility",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "depth": 5,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612227218481,
      "hashes": {
        "sha256": "/lSrm4/Y4CWDGsVRZnFSbiP3YRNivFR7TQARVtMXQQU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "mjaE56SnX3A3C4soy48lKevOtjLxGqv4gxavCg0tsxfSCMhvVlzbYMofAX23K7Aq8Tg4cF3bf8HIVMkeNsLMAA"
        }
      },
      "unsigned": {
        "age": 447
      },
      "event_id": "$xCtSXTydyn3uatfvyl_C8-Iq2wxgRpC5bSxNc9191-g",
      "user_id": "@alice:hs1",
      "age": 447
    },
    {
      "auth_events": [
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY"
      ],
      "prev_events": [
        "$Z4S9hTbiY8JbSSZK711oQ2Kr2TS4mCfiNSqAeZYdBLY"
      ],
      "type": "m.room.member",
      "room_id": "!bBAwYynvirrDqcGXRJ:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "depth": 2,
      "prev_state": [],
      "state_key": "@alice:hs1",
      "origin": "hs1",
      "origin_server_ts": 1612227218374,
      "hashes": {
        "sha256": "Ijm11zXRu0bHwFUSDCULX878nJtVXxSYUWbdYcMSBQs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_BXDh": "guhM3fw+cXOOQ++Qb/wmZJkednj7XnTDqMx84daq3BO8FGznV4kYrpb1PWKFhsCRX3IULSCNq/+0Q0ysdM9FBA"
        }
      },
      "unsigned": {
        "age": 554
      },
      "event_id": "$-XqUjBEMDviU9U9l9pjXsW8uyM98EWFQjH0lxGLHXm8",
      "user_id": "@alice:hs1",
      "age": 554
    }
  ]
}
```

</details>

Reply:
Looks like I need to update to make the `depth` match at where I am inserting at. Make the depth be the same as the `prev_event` defined

Reply:
Now that I got `depth` is inherited from the `prev_event` you're inserting next to, the ordering is looking correct from `/messages` 🎉 


`A <- 1 <- 2 <- 3 <- B <- C`




[`GET /_matrix/client/r0/rooms/{roomId}/messages`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-messages)

<details>

<summary>Response:</summary>

```json
{
  "chunk": [
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$_D6pi0LvPled67aB6nWMyScbAXAQbk_9WPHSqZ6xw1Y",
        "$_ZV6nkZELUiI3sULERpWWVvGYUxybhiabstSeIsJQUQ"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message *",
        "msgtype": "m.text"
      },
      "depth": 8,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612262669217,
      "hashes": {
        "sha256": "fLO/8azm/YxYVOB+1DpFljrKpiHVcYmUToL/DuHXz6Y"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "PYFtDAzK1pI8XTi4WXWHy0YeFJ+LCMHUSjLuSG8MXDkzeRKTaV9VVMLKvmqTiBq5PGD0deiqtF/JA28l2eSnBQ"
        }
      },
      "unsigned": {
        "age": 54
      },
      "event_id": "$D213YOhnwuNYdCeKjnvs_XEAz7qEqaROweAG7a1ivBQ",
      "user_id": "@alice:hs1",
      "age": 54
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$Mg-Lx_gu-X1tavThwpHYUUOoR4ggwlhoSBd0EyPb3o8"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message C",
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612262669072,
      "hashes": {
        "sha256": "lVHaU/JVMh1sNokxx0RsoCMFxmqrv+mnUAjZh7cKqK4"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "tfvc2pv/5l+37mD1nKk0zp6CUBM934B6XY/2FaDQxwsDrvBa0iPerbdd4VBIy/BR5u/YL2PuN+m8tL8yIGbdCA"
        }
      },
      "unsigned": {
        "age": 199
      },
      "event_id": "$_D6pi0LvPled67aB6nWMyScbAXAQbk_9WPHSqZ6xw1Y",
      "user_id": "@alice:hs1",
      "age": 199
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$_VDUgOO1K8nFmA44fl_xq2C20bZ14hCKSgvvMEhPHqw"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message B",
        "msgtype": "m.text"
      },
      "depth": 6,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612262669024,
      "hashes": {
        "sha256": "+1xrtQEtfhew7hqb0pSbSPRJjl82yR+TEdY1d4AtO5Q"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "xn9wpEiLdA4/hGRrRTjRY9WE99YzJgCx2fTbR0+om1tJTFtVr4n8tzac8wxN6eN7XyzLwrmWTBda01og1dvWDg"
        }
      },
      "unsigned": {
        "age": 247
      },
      "event_id": "$Mg-Lx_gu-X1tavThwpHYUUOoR4ggwlhoSBd0EyPb3o8",
      "user_id": "@alice:hs1",
      "age": 247
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$LiFY16IYtjFhj0U0rK4RZ0zTluFOI6tqDkjj6-jbYw4"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 3",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin_server_ts": 1612262669011,
      "origin": "hs1",
      "hashes": {
        "sha256": "aESfb9LOfaJsPTqsWq64I2KCPXov8dgxo3ToinD9Jz4"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "tPizbni/DUsrjHAO9HK8t6HAvlRmUXqHWaddmOwW2c/9cieOsJiQeciFmH+Jvzxj0GjiKbZT5YZzG5dCe+WGAA"
        }
      },
      "unsigned": {
        "age": 92
      },
      "event_id": "$_ZV6nkZELUiI3sULERpWWVvGYUxybhiabstSeIsJQUQ",
      "user_id": "@alice:hs1",
      "age": 92
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$RDvxWDNy_GLRrIlYsF5RdBt-FVDXEOqBsMp-J1JMb-Q"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 2",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin_server_ts": 1612262669010,
      "origin": "hs1",
      "hashes": {
        "sha256": "HuK231UFGii7NPRL7Jq1nvR56UZ8QvkVf0sveusI39E"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "L8g20O2yHE4ESmLB+1y9l9JZzm4KajsSnjNPawdShNar9mbFkvl4iadFA5T0ysy2JCMj1Dk3gZP0eO9vcQSLCw"
        }
      },
      "unsigned": {
        "age": 125
      },
      "event_id": "$LiFY16IYtjFhj0U0rK4RZ0zTluFOI6tqDkjj6-jbYw4",
      "user_id": "@alice:hs1",
      "age": 125
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$_VDUgOO1K8nFmA44fl_xq2C20bZ14hCKSgvvMEhPHqw"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 1",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin_server_ts": 1612262669009,
      "origin": "hs1",
      "hashes": {
        "sha256": "AgFLp0Rf2/ApmlvsUTA3rlC7hZ2uBR5HCDz91MO+Rn8"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "lWgjf19AurIVENBtqq2H0U/aMkbLQd/l05VlsFUAvJFdnxHr4pZKJWLj0A4lDfwD1XEKH3PtVxwyFHfihu7sBg"
        }
      },
      "unsigned": {
        "age": 150
      },
      "event_id": "$RDvxWDNy_GLRrIlYsF5RdBt-FVDXEOqBsMp-J1JMb-Q",
      "user_id": "@alice:hs1",
      "age": 150
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$hrcm1rZtft5669sh90mEZtGOh74QXrxL-MKXS7BA9pw"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message A",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612262668937,
      "hashes": {
        "sha256": "QHg/EIbIWYqU44LAlfpeZZtAytfuYhLpUuIDY6rS1IA"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "5Z8lJp+SHlL1YnMwa+FCTZZdQh/Y2JpfcuTkwRFEkCplKdi6ubzVn1DYRogw/R1jOZhwX/lRXmD/bkYrAyczCQ"
        }
      },
      "unsigned": {
        "age": 334
      },
      "event_id": "$_VDUgOO1K8nFmA44fl_xq2C20bZ14hCKSgvvMEhPHqw",
      "user_id": "@alice:hs1",
      "age": 334
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$4L3Fq0g3RL03QVTnqpKNsgKcc6LTXYNkXBBVQt_lzCM"
      ],
      "type": "m.room.history_visibility",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "depth": 4,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668879,
      "hashes": {
        "sha256": "AUGOXbgTTpFtTpMHYOVQ+mDd3d1MKcAr4IclIhFEuqs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "aU04UKDIGhhdoJ9a9VIFMOauCY/B21LlhKi+W+XnypE5kZEKMetMLDrhAcfM865fReqgsdAGGQQrtkrM2NOPCA"
        }
      },
      "unsigned": {
        "age": 392
      },
      "event_id": "$hrcm1rZtft5669sh90mEZtGOh74QXrxL-MKXS7BA9pw",
      "user_id": "@alice:hs1",
      "age": 392
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o"
      ],
      "type": "m.room.join_rules",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "join_rule": "public"
      },
      "depth": 3,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668848,
      "hashes": {
        "sha256": "Iq9MwVtG2UIFyUntENL1EA2AX42hIcs9dEtekvBxvvQ"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "ze57BsakFHpg5JMFHM5rAjI76ZQd32gYPXD+F0EEVpJJgV61QPnHZQmKg3vpz1bl4JeMNCVB408QmhldGGnZDg"
        }
      },
      "unsigned": {
        "age": 423
      },
      "event_id": "$4L3Fq0g3RL03QVTnqpKNsgKcc6LTXYNkXBBVQt_lzCM",
      "user_id": "@alice:hs1",
      "age": 423
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "type": "m.room.power_levels",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "users": {
          "@alice:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "depth": 2,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668816,
      "hashes": {
        "sha256": "0FTuymkzxQDhuEvp1a9/ZM3p/cfZGFzaxyQzQp5XHtw"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "WBLm5HT21dqJtA3LrKjbxITQLMjN/Vki5QNmL7zf0HVRhL9z2bneh07YQOD+6k2YLUKAbj1MDruRI9Rq+r4wCQ"
        }
      },
      "unsigned": {
        "age": 455
      },
      "event_id": "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
      "user_id": "@alice:hs1",
      "age": 455
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc"
      ],
      "prev_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc"
      ],
      "type": "m.room.member",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "@alice:hs1",
      "origin": "hs1",
      "origin_server_ts": 1612262668775,
      "hashes": {
        "sha256": "R8X4VkpZlMuqnfl9Qxa4fxw9bBPfLnKo+yedNeCLTfo"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "23vpgg2BKsiW26BoNaTlGWuCB3R01Yz/S+n+8WKomj3xEuicVHaJvZcc5V2gwDCD5mJJ4iI7V0J+n+OjIMb5CA"
        }
      },
      "unsigned": {
        "age": 496
      },
      "event_id": "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y",
      "user_id": "@alice:hs1",
      "age": 496
    },
    {
      "auth_events": [],
      "prev_events": [],
      "type": "m.room.create",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "room_version": "6",
        "creator": "@alice:hs1"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668717,
      "hashes": {
        "sha256": "2oBNBE2BQeexnf2Brb3u94OTb4b9MWzcaBoAQB45NHU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "mtXhbuVtBn6vcbQG93wi4kQv0n7omtb8YRotL9Rf9tBqSUIW71Nb9Bja0XtoWSOfRJPA9+5fu3CEEOSfGARzCw"
        }
      },
      "unsigned": {
        "age": 554
      },
      "event_id": "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
      "user_id": "@alice:hs1",
      "age": 554
    }
  ],
  "start": "s13_0_0_0_0_0_0_0_0",
  "end": "t1-1_0_0_0_0_0_0_0_0"
}
```

</details>


![](https://user-images.githubusercontent.com/558581/106590309-92b3b680-6512-11eb-8317-f4e6a1b9362e.png)






---




[`GET /_matrix/client/r0/rooms/{roomId}/context/{eventId}`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-context-eventid)
<details>

<summary>Response:</summary>

```json
{
  "events_before": [
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$Mg-Lx_gu-X1tavThwpHYUUOoR4ggwlhoSBd0EyPb3o8"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message C",
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612262669072,
      "hashes": {
        "sha256": "lVHaU/JVMh1sNokxx0RsoCMFxmqrv+mnUAjZh7cKqK4"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "tfvc2pv/5l+37mD1nKk0zp6CUBM934B6XY/2FaDQxwsDrvBa0iPerbdd4VBIy/BR5u/YL2PuN+m8tL8yIGbdCA"
        }
      },
      "unsigned": {
        "age": 230
      },
      "event_id": "$_D6pi0LvPled67aB6nWMyScbAXAQbk_9WPHSqZ6xw1Y",
      "user_id": "@alice:hs1",
      "age": 230
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$_VDUgOO1K8nFmA44fl_xq2C20bZ14hCKSgvvMEhPHqw"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message B",
        "msgtype": "m.text"
      },
      "depth": 6,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612262669024,
      "hashes": {
        "sha256": "+1xrtQEtfhew7hqb0pSbSPRJjl82yR+TEdY1d4AtO5Q"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "xn9wpEiLdA4/hGRrRTjRY9WE99YzJgCx2fTbR0+om1tJTFtVr4n8tzac8wxN6eN7XyzLwrmWTBda01og1dvWDg"
        }
      },
      "unsigned": {
        "age": 278
      },
      "event_id": "$Mg-Lx_gu-X1tavThwpHYUUOoR4ggwlhoSBd0EyPb3o8",
      "user_id": "@alice:hs1",
      "age": 278
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$LiFY16IYtjFhj0U0rK4RZ0zTluFOI6tqDkjj6-jbYw4"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 3",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin_server_ts": 1612262669011,
      "origin": "hs1",
      "hashes": {
        "sha256": "aESfb9LOfaJsPTqsWq64I2KCPXov8dgxo3ToinD9Jz4"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "tPizbni/DUsrjHAO9HK8t6HAvlRmUXqHWaddmOwW2c/9cieOsJiQeciFmH+Jvzxj0GjiKbZT5YZzG5dCe+WGAA"
        }
      },
      "unsigned": {
        "age": 123
      },
      "event_id": "$_ZV6nkZELUiI3sULERpWWVvGYUxybhiabstSeIsJQUQ",
      "user_id": "@alice:hs1",
      "age": 123
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$RDvxWDNy_GLRrIlYsF5RdBt-FVDXEOqBsMp-J1JMb-Q"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 2",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin_server_ts": 1612262669010,
      "origin": "hs1",
      "hashes": {
        "sha256": "HuK231UFGii7NPRL7Jq1nvR56UZ8QvkVf0sveusI39E"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "L8g20O2yHE4ESmLB+1y9l9JZzm4KajsSnjNPawdShNar9mbFkvl4iadFA5T0ysy2JCMj1Dk3gZP0eO9vcQSLCw"
        }
      },
      "unsigned": {
        "age": 156
      },
      "event_id": "$LiFY16IYtjFhj0U0rK4RZ0zTluFOI6tqDkjj6-jbYw4",
      "user_id": "@alice:hs1",
      "age": 156
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$_VDUgOO1K8nFmA44fl_xq2C20bZ14hCKSgvvMEhPHqw"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 1",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin_server_ts": 1612262669009,
      "origin": "hs1",
      "hashes": {
        "sha256": "AgFLp0Rf2/ApmlvsUTA3rlC7hZ2uBR5HCDz91MO+Rn8"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "lWgjf19AurIVENBtqq2H0U/aMkbLQd/l05VlsFUAvJFdnxHr4pZKJWLj0A4lDfwD1XEKH3PtVxwyFHfihu7sBg"
        }
      },
      "unsigned": {
        "age": 181
      },
      "event_id": "$RDvxWDNy_GLRrIlYsF5RdBt-FVDXEOqBsMp-J1JMb-Q",
      "user_id": "@alice:hs1",
      "age": 181
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$hrcm1rZtft5669sh90mEZtGOh74QXrxL-MKXS7BA9pw"
      ],
      "type": "m.room.message",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message A",
        "msgtype": "m.text"
      },
      "depth": 5,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1612262668937,
      "hashes": {
        "sha256": "QHg/EIbIWYqU44LAlfpeZZtAytfuYhLpUuIDY6rS1IA"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "5Z8lJp+SHlL1YnMwa+FCTZZdQh/Y2JpfcuTkwRFEkCplKdi6ubzVn1DYRogw/R1jOZhwX/lRXmD/bkYrAyczCQ"
        }
      },
      "unsigned": {
        "age": 365
      },
      "event_id": "$_VDUgOO1K8nFmA44fl_xq2C20bZ14hCKSgvvMEhPHqw",
      "user_id": "@alice:hs1",
      "age": 365
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$4L3Fq0g3RL03QVTnqpKNsgKcc6LTXYNkXBBVQt_lzCM"
      ],
      "type": "m.room.history_visibility",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "depth": 4,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668879,
      "hashes": {
        "sha256": "AUGOXbgTTpFtTpMHYOVQ+mDd3d1MKcAr4IclIhFEuqs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "aU04UKDIGhhdoJ9a9VIFMOauCY/B21LlhKi+W+XnypE5kZEKMetMLDrhAcfM865fReqgsdAGGQQrtkrM2NOPCA"
        }
      },
      "unsigned": {
        "age": 423
      },
      "event_id": "$hrcm1rZtft5669sh90mEZtGOh74QXrxL-MKXS7BA9pw",
      "user_id": "@alice:hs1",
      "age": 423
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o"
      ],
      "type": "m.room.join_rules",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "join_rule": "public"
      },
      "depth": 3,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668848,
      "hashes": {
        "sha256": "Iq9MwVtG2UIFyUntENL1EA2AX42hIcs9dEtekvBxvvQ"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "ze57BsakFHpg5JMFHM5rAjI76ZQd32gYPXD+F0EEVpJJgV61QPnHZQmKg3vpz1bl4JeMNCVB408QmhldGGnZDg"
        }
      },
      "unsigned": {
        "age": 454
      },
      "event_id": "$4L3Fq0g3RL03QVTnqpKNsgKcc6LTXYNkXBBVQt_lzCM",
      "user_id": "@alice:hs1",
      "age": 454
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "type": "m.room.power_levels",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "users": {
          "@alice:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "depth": 2,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668816,
      "hashes": {
        "sha256": "0FTuymkzxQDhuEvp1a9/ZM3p/cfZGFzaxyQzQp5XHtw"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "WBLm5HT21dqJtA3LrKjbxITQLMjN/Vki5QNmL7zf0HVRhL9z2bneh07YQOD+6k2YLUKAbj1MDruRI9Rq+r4wCQ"
        }
      },
      "unsigned": {
        "age": 486
      },
      "event_id": "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
      "user_id": "@alice:hs1",
      "age": 486
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc"
      ],
      "prev_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc"
      ],
      "type": "m.room.member",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "@alice:hs1",
      "origin": "hs1",
      "origin_server_ts": 1612262668775,
      "hashes": {
        "sha256": "R8X4VkpZlMuqnfl9Qxa4fxw9bBPfLnKo+yedNeCLTfo"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "23vpgg2BKsiW26BoNaTlGWuCB3R01Yz/S+n+8WKomj3xEuicVHaJvZcc5V2gwDCD5mJJ4iI7V0J+n+OjIMb5CA"
        }
      },
      "unsigned": {
        "age": 527
      },
      "event_id": "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y",
      "user_id": "@alice:hs1",
      "age": 527
    },
    {
      "auth_events": [],
      "prev_events": [],
      "type": "m.room.create",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "room_version": "6",
        "creator": "@alice:hs1"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668717,
      "hashes": {
        "sha256": "2oBNBE2BQeexnf2Brb3u94OTb4b9MWzcaBoAQB45NHU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "mtXhbuVtBn6vcbQG93wi4kQv0n7omtb8YRotL9Rf9tBqSUIW71Nb9Bja0XtoWSOfRJPA9+5fu3CEEOSfGARzCw"
        }
      },
      "unsigned": {
        "age": 585
      },
      "event_id": "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
      "user_id": "@alice:hs1",
      "age": 585
    }
  ],
  "events_after": [],
  "start": "t1-1_0_0_0_0_0_0_0_0",
  "end": "t8-13_0_0_0_0_0_0_0_0",
  "event": {
    "auth_events": [
      "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
      "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
      "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
    ],
    "prev_events": [
      "$_D6pi0LvPled67aB6nWMyScbAXAQbk_9WPHSqZ6xw1Y",
      "$_ZV6nkZELUiI3sULERpWWVvGYUxybhiabstSeIsJQUQ"
    ],
    "type": "m.room.message",
    "room_id": "!okWkdbRLVcCMKUgGow:hs1",
    "sender": "@alice:hs1",
    "content": {
      "body": "Message *",
      "msgtype": "m.text"
    },
    "depth": 8,
    "prev_state": [],
    "origin": "hs1",
    "origin_server_ts": 1612262669217,
    "hashes": {
      "sha256": "fLO/8azm/YxYVOB+1DpFljrKpiHVcYmUToL/DuHXz6Y"
    },
    "signatures": {
      "hs1": {
        "ed25519:a_KTOx": "PYFtDAzK1pI8XTi4WXWHy0YeFJ+LCMHUSjLuSG8MXDkzeRKTaV9VVMLKvmqTiBq5PGD0deiqtF/JA28l2eSnBQ"
      }
    },
    "unsigned": {
      "age": 85
    },
    "event_id": "$D213YOhnwuNYdCeKjnvs_XEAz7qEqaROweAG7a1ivBQ",
    "user_id": "@alice:hs1",
    "age": 85
  },
  "state": [
    {
      "auth_events": [],
      "prev_events": [],
      "type": "m.room.create",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "room_version": "6",
        "creator": "@alice:hs1"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668717,
      "hashes": {
        "sha256": "2oBNBE2BQeexnf2Brb3u94OTb4b9MWzcaBoAQB45NHU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "mtXhbuVtBn6vcbQG93wi4kQv0n7omtb8YRotL9Rf9tBqSUIW71Nb9Bja0XtoWSOfRJPA9+5fu3CEEOSfGARzCw"
        }
      },
      "unsigned": {
        "age": 585
      },
      "event_id": "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
      "user_id": "@alice:hs1",
      "age": 585
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "type": "m.room.power_levels",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "users": {
          "@alice:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "depth": 2,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668816,
      "hashes": {
        "sha256": "0FTuymkzxQDhuEvp1a9/ZM3p/cfZGFzaxyQzQp5XHtw"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "WBLm5HT21dqJtA3LrKjbxITQLMjN/Vki5QNmL7zf0HVRhL9z2bneh07YQOD+6k2YLUKAbj1MDruRI9Rq+r4wCQ"
        }
      },
      "unsigned": {
        "age": 486
      },
      "event_id": "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
      "user_id": "@alice:hs1",
      "age": 486
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o"
      ],
      "type": "m.room.join_rules",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "join_rule": "public"
      },
      "depth": 3,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668848,
      "hashes": {
        "sha256": "Iq9MwVtG2UIFyUntENL1EA2AX42hIcs9dEtekvBxvvQ"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "ze57BsakFHpg5JMFHM5rAjI76ZQd32gYPXD+F0EEVpJJgV61QPnHZQmKg3vpz1bl4JeMNCVB408QmhldGGnZDg"
        }
      },
      "unsigned": {
        "age": 454
      },
      "event_id": "$4L3Fq0g3RL03QVTnqpKNsgKcc6LTXYNkXBBVQt_lzCM",
      "user_id": "@alice:hs1",
      "age": 454
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc",
        "$u_yI1aE1-O2fUhO2692FU7bPxozEW30a4wB3SdYxu4o",
        "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y"
      ],
      "prev_events": [
        "$4L3Fq0g3RL03QVTnqpKNsgKcc6LTXYNkXBBVQt_lzCM"
      ],
      "type": "m.room.history_visibility",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "depth": 4,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1612262668879,
      "hashes": {
        "sha256": "AUGOXbgTTpFtTpMHYOVQ+mDd3d1MKcAr4IclIhFEuqs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "aU04UKDIGhhdoJ9a9VIFMOauCY/B21LlhKi+W+XnypE5kZEKMetMLDrhAcfM865fReqgsdAGGQQrtkrM2NOPCA"
        }
      },
      "unsigned": {
        "age": 423
      },
      "event_id": "$hrcm1rZtft5669sh90mEZtGOh74QXrxL-MKXS7BA9pw",
      "user_id": "@alice:hs1",
      "age": 423
    },
    {
      "auth_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc"
      ],
      "prev_events": [
        "$P8fznaoxJu4YRpalUxPedYWvHs7tqYRliXcpxfskGfc"
      ],
      "type": "m.room.member",
      "room_id": "!okWkdbRLVcCMKUgGow:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "@alice:hs1",
      "origin": "hs1",
      "origin_server_ts": 1612262668775,
      "hashes": {
        "sha256": "R8X4VkpZlMuqnfl9Qxa4fxw9bBPfLnKo+yedNeCLTfo"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_KTOx": "23vpgg2BKsiW26BoNaTlGWuCB3R01Yz/S+n+8WKomj3xEuicVHaJvZcc5V2gwDCD5mJJ4iI7V0J+n+OjIMb5CA"
        }
      },
      "unsigned": {
        "age": 527
      },
      "event_id": "$j9U-S7ntLXpshVGVyPErHQuxWo_En7tx5ANPo8FXH3Y",
      "user_id": "@alice:hs1",
      "age": 527
    }
  ]
}
```

</details>